lwt-core/cohttp_lwt.ml

```diff
@@ -206,32 +206,29 @@ module Make_client
   let callv ?(ctx=default_ctx) uri reqs =
     Net.connect_uri ~ctx uri >>= fun (conn, ic, oc) ->
     (* Serialise the requests out to the wire *)
-    Lwt_stream.fold_s (fun (req,body) meths ->
+    let meth_stream = Lwt_stream.map_s (fun (req,body) ->
       Request.write (fun writer ->
         Cohttp_lwt_body.write_body (Request.write_body writer) body
       ) req oc >>= fun () ->
-      return ((Request.meth req)::meths)
-    ) reqs [] >>= fun meths ->
+      return (Request.meth req)
+    ) reqs in
     (* Read the responses. For each response, ensure that the previous
        response has consumed the body before continuing to the next
        response because HTTP/1.1-pipelining cannot be interleaved. *)
```

Comment:
On 06/29/2015 04:08 PM, David Sheets wrote:
> In lwt-core/cohttp_lwt.ml <https://github.com/mirage/ocaml-cohttp/pull/379#discussion_r33461134>:
> 
>>        Request.write (fun writer ->
>>          Cohttp_lwt_body.write_body (Request.write_body writer) body
>>        ) req oc >>= fun () ->
>> -      return ((Request.meth req)::meths)
>> -    ) reqs [] >>= fun meths ->
>> +      return (Request.meth req)
>> +    ) reqs in
>>      (* Read the responses. For each response, ensure that the previous
>>         response has consumed the body before continuing to the next
>>         response because HTTP/1.1-pipelining cannot be interleaved. *)
> 
> This comment seems a little wrong. Does it mean HTTP/1.1 pipelining cannot be /out-of-order/? It is called /pipelining/...

The server is free to execute the requests in whatever order as long as it ensures that the replies are sent in the same order
as the requests: https://tools.ietf.org/html/rfc7230#page-54

There is nothing special in HTTP/1.1 pipelining that would allow clients to match responses with requests[1],
hence the pipelineing blacklists you see in tools like curl to deal with buggy servers that send replies out-of-order.

FWIW Firefox has pipelineing disabled by default, and chromium is about to remove it too: https://bugzilla.mozilla.org/show_bug.cgi?id=264354
Still I think it'd be useful for cohttp in situations where you know the server works correctly, in other situations you can still use persistence by specifying pipeline depth of 0.

[1] there was (a now expired) draft to add a header Assoc-Req: https://tools.ietf.org/html/rfc7230#page-54